### PR TITLE
WIP : Opentsdb alias fix

### PR DESCRIPTION
**Pull request by utkarshcmu**
This PR requires Opentsdb 2.2 to work.

### CODE REVIEW

**Comment by torkelo** [comment]
What do you mean it requires it? will it break or what will happen for users using older versions?

File: public/app/plugins/datasource/opentsdb/datasource.js [hunk 1]
```diff
@@ -293,11 +294,12 @@ function (angular, _, dateMath) {
...
         return _.findIndex(options.targets, function(target) {
-          return target.metric === metricData.metric &&
-            _.all(target.tags, function(tagV, tagK) {
-            interpolatedTagValue = templateSrv.replace(tagV, options.scopedVars);
-            return metricData.tags[tagK] === interpolatedTagValue || interpolatedTagValue === "*";
-          });
+          return target.metric === metricData.query.metric &&
+          _.all(target.tags, function(tagV, tagK) {
+              interpolatedTagValue = templateSrv.replace(tagV, options.scopedVars);
+              return metricData.query.tags[tagK] === "literal_or("+interpolatedTagValue+")" || interpolatedTagValue === "*";
```
Comment (torkelo) [inline public/app/plugins/datasource/opentsdb/datasource.js]: what is this? literal_or ? you do not seem to use the showQuery feature?

File: public/app/plugins/datasource/opentsdb/datasource.js [hunk 2]
```diff
@@ -63,7 +63,8 @@ function (angular, _, dateMath) {
     OpenTSDBDatasource.prototype.performTimeSeriesQuery = function(queries, start, end) {
       var reqBody = {
         start: start,
-        queries: queries
+        queries: queries,
+        showQuery: true
```
Comment (utkarshcmu) [inline public/app/plugins/datasource/opentsdb/datasource.js]: @torkelo I am using showQuery feature on this line here.

File: public/app/plugins/datasource/opentsdb/datasource.js (hunk 1 above)
Comment (utkarshcmu) [inline public/app/plugins/datasource/opentsdb/datasource.js]: So, in 2.2.0, Opentsdb has deprecated tags and introduced filters of 6 types. One of the type is "literal_or" (equivalent to piped tag values).

For example, when we send a tag key/value pair as following to the Opentsdb 2.2.0:
```env:prod-1|prod-2|prod-3```
Opentsdb converts this tag key/value into filters and returns the following:
```literal_or(env:prod-1|prod-2|prod-3)```

So, now to match the query, we need to match the request values to response values.

**Comment by utkarshcmu** [comment]
@torkelo I think we might need to make two datasources for Opentsdb as well like InfluxDB, so that Grafana is compatible with all Opentsdb versions and people do not have to upgrade Opentsdb if they are upgrading Grafana:

1. Until Opentsdb 2.1.3
2. Above Opentsdb 2.1.3

What do you think?

**Comment by torkelo** [comment]
maybe, but still seems like you are not using the showQuery feature? With that each series should have a query id of some sort to match it to the query, right?

**Comment by torkelo** [comment]
I mean the whole point of showQuery was to to have the OpenTSDB response show which series came from which query, maybe via a query index?

**Comment by utkarshcmu** [comment]
@torkelo I don't see any queryId or index in the response from Opentsdb 2.2.0. Here is an example:

Request:
http://my_opentsdb_url/api/query

Parameters:
```
{
    "start": 1447786535237,
    "queries": [
        {
            "metric": "load.load.shortterm",
            "aggregator": "sum",
            "tags": {
                "app": "activity|aggregator"
            }
        },
        {
            "metric": "load.load.shortterm",
            "aggregator": "avg",
            "downsample": "1s-avg",
            "tags": {
                "app": "activity|aggregator"
            }
        }
    ],
    "showQuery": true
}
```

Now here is the response which we get from Opentsdb 2.2.0. The above request should return an array of 4 series because of 2 tags in each query which it does. Also each series has a query field within it:

Response JSON:
```
[
    {
        "metric": "load.load.shortterm",
        "tags": {
            "instance_type": "ec2",
            "app": "activity",
            "data_source": "collectd",
            "env": "p1-np"
        },
        "aggregateTags": [
            "instance_id",
            "ip"
        ],
        "query": {
            "aggregator": "sum",
            "metric": "load.load.shortterm",
            "tsuids": null,
            "downsample": null,
            "rate": false,
            "filters": [
                {
                    "tagk": "app",
                    "filter": "activity|aggregator",
                    "group_by": true,
                    "type": "literal_or"
                }
            ],
            "rateOptions": null,
            "tags": {
                "app": "literal_or(activity|aggregator)"
            }
        },
        "dps": {
            .....
        }
    },
    {
        "metric": "load.load.shortterm",
        "tags": {
            "instance_type": "ec2",
            "app": "aggregator",
            "data_source": "collectd",
            "env": "p1-np"
        },
        "aggregateTags": [
            "instance_id",
            "ip"
        ],
        "query": {
            "aggregator": "sum",
            "metric": "load.load.shortterm",
            "tsuids": null,
            "downsample": null,
            "rate": false,
            "filters": [
                {
                    "tagk": "app",
                    "filter": "activity|aggregator",
                    "group_by": true,
                    "type": "literal_or"
                }
            ],
            "rateOptions": null,
            "tags": {
                "app": "literal_or(activity|aggregator)"
            }
        },
        "dps": {......
        }
    },
    {
        "metric": "load.load.shortterm",
        "tags": {
            "instance_type": "ec2",
            "app": "activity",
            "data_source": "collectd",
            "env": "p1-np"
        },
        "aggregateTags": [
            "instance_id",
            "ip"
        ],
        "query": {
            "aggregator": "avg",
            "metric": "load.load.shortterm",
            "tsuids": null,
            "downsample": "1s-avg",
            "rate": false,
            "filters": [
                {
                    "tagk": "app",
                    "filter": "activity|aggregator",
                    "group_by": true,
                    "type": "literal_or"
                }
            ],
            "rateOptions": null,
            "tags": {
                "app": "literal_or(activity|aggregator)"
            }
        },
        "dps": {.......
        }
    },
    {
        "metric": "load.load.shortterm",
        "tags": {
            "instance_type": "ec2",
            "app": "aggregator",
            "data_source": "collectd",
            "env": "p1-np"
        },
        "aggregateTags": [
            "instance_id",
            "ip"
        ],
        "query": {
            "aggregator": "avg",
            "metric": "load.load.shortterm",
            "tsuids": null,
            "downsample": "1s-avg",
            "rate": false,
            "filters": [
                {
                    "tagk": "app",
                    "filter": "activity|aggregator",
                    "group_by": true,
                    "type": "literal_or"
                }
            ],
            "rateOptions": null,
            "tags": {
                "app": "literal_or(activity|aggregator)"
            }
        },
        "dps": {.....
        }
    }
]
```

**Comment by utkarshcmu** [comment]
The only way I can think right now to match a response to a query field is to compare either all fields or at least the major ones and that is what I implemented in this PR.

**Comment by torkelo** [comment]
This should be reported to OpenTSDB team, that makes the showQuery broken/pointless, they should have added a query index / id.

**Comment by utkarshcmu** [comment]
@torkelo Closing this PR as it breaks Opentsdb 2.1.x support. Will open another one with Opentsdb 2.2.0 support with other features as well.

**Comment by utkarshcmu** [comment]
Closing this PR permanently as https://github.com/OpenTSDB/opentsdb/issues/646 will not be released in 2.2.0.

**Pull request closed without merging**